projects/laji/src/app/+project-form/form/document-form/document-form.facade.ts

```diff
@@ -509,11 +509,11 @@ export class DocumentFormFacade {
   }
 
   private getAnnotations(documentID: string, page = 1, results = []): Observable<Annotation[]> {
-    if (this.annotationCache[documentID]) {
-      return this.annotationCache[documentID];
+    if (this.annotationCache[this.getAnnotationCacheKey(documentID, page)]) {
```

Comment:
This long line is repeated so many times that I think it could be assigned to variable `const cacheKey = this.getAnnotationCacheKey(documentID, page)`

Reply:
Now that I looked it more, the cache seems to work little weird. When page = 1, it saves all the whole annotation list to cache, when page = 2, it saves all the annotations from page 2 onwards to cache and so on... So the cache seems unnecessary if the page is something else than 1

Reply:
One suggestion for fixing this is to rename `getAnnotations` to `getAnnotationsFromPage` and make a new `getAnnotations(documentID)` function that handles the caching and calls `getAnnotationsFromPage(documentID, 1)`

Reply:
I removed the `this.getAnnotationCacheKey` altogether, since the cache key can be created in a single line.

> Now that I looked it more, the cache seems to work little weird. When page = 1, it saves the whole annotation list to cache, when page = 2, it saves all the annotations from page 2 onwards to cache and so on... So the cache seems unnecessary if the page is something else than 1

Hmm... I don't find the cache unnecessary. The `getAnnotations()` works recursively, so it will first get the first page, and then get the consecutive pages if needed and all the pages are cached.

Reply:
Yeah it works recursively so the the cache will look like this
`annotationCache[doc + 1] = Observable<annotations from page 1 to 100>`
`annotationCache[doc + 2] = Observable<annotations from page 2 to 100>`
`annotationCache[doc + 3] = Observable<annotations from page 3 to 100>`
`...`
So in this scenario it will cache annotations that are in page 100 100 times...

Reply:
> So in this scenario it will cache annotations that are in page 100 100 times...

I think it fetches the page 100 only once. I see it like this:

```
getAnnotations(doc, 1) = annotationCache[doc, 1] + Observable<annotations from page 1> + getAnnotations(doc, 2) 
getAnnotations(doc, 2) = annotationCache[doc, 2] + Observable<annotations from page 1> + getAnnotations(doc, 3) 
...
getAnnotations(doc, 99) = annotationCache[doc, 99] + Observable<annotations from page 1> + getAnnotations(doc, 100) 
getAnnotations(doc, 100) = annotationCache[doc, 100]
```

The results are stored to memory in each call though (`[...results, ...result.results]`) before the call stack is complete thought, which isn't optimal... Realistically that isn't a problem. I wonder if we have some page iteration utility somewhere for cases like this.